### PR TITLE
[Flax] Big FlaxBert Refactor

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR does a major refactor of FlaxBert in Transformers, notably:
- Costume LayerNorm and Embedding layers are replaced by the official ones. This should significantly reduce maintenance cost at the expense of two more general lines in the conversion script.
- A couple of bugs are fixed, *e.g.*, BERT uses the *non-approximated* GELU and not the fast one -> this fixes some minor differences when comparing PyTorchBERT vs FlaxBERT
- Weight Tying is added, which should be done for, *e.g.* `FlaxBertForMaskedLM`
- Weights can now also be converted the other way around Flax => PyTorch

Sorry for putting quite a lot of things into one PR, but there are very much intertwined here.

Also, I will have to re-upload some flax weights so that they correspond to the new weight structure (see [here](https://github.com/huggingface/transformers/pull/10977))

@avital @marcvanzee, I had an issue when saving/loading flax weights for which I opened an issue [here](https://github.com/google/flax/issues/1261). At the moment, I solve it by manually transforming every numpy array into a jax DeviceArray, see [here](https://github.com/huggingface/transformers/pull/11364/files#r617853008) - not sure if there is a better solution.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/modeling_flax_utils.py [hunk 1]
```diff
@@ -46,7 +46,7 @@
 
 
 ACT2FN = {
-    "gelu": nn.gelu,
+    "gelu": partial(nn.gelu, approximate=False),
```
Comment (patrickvonplaten) [inline src/transformers/modeling_flax_utils.py:49]: This is the correct gelu function for BERT -> error tolerance can be decreased to 1e-3 thanks to this correction

File: src/transformers/modeling_flax_utils.py [hunk 2]
```diff
@@ -100,9 +100,22 @@ def __init__(
         self._required_params = set(flatten_dict(unfreeze(random_params)).keys())
         self.params = random_params
 
+        # tie input and output word embeddings
+        # if relevant
```
Comment (patrickvonplaten) [inline src/transformers/modeling_flax_utils.py:104]:
```suggestion
```

File: src/transformers/modeling_flax_utils.py [hunk 3]
```diff
@@ -100,9 +100,21 @@ def __init__(
         self._required_params = set(flatten_dict(unfreeze(random_params)).keys())
         self.params = random_params
 
+        # tie input and output word embeddings
```
Comment (patrickvonplaten) [inline src/transformers/modeling_flax_utils.py:103]:
```suggestion
        # tie input and output word embeddings if model has output word embeddings 
        and ``self.config.tie_word_embedding is True``
```

File: src/transformers/modeling_flax_utils.py [hunk 4]
```diff
@@ -100,9 +100,21 @@ def __init__(
...
+
     def init_weights(self, rng: jax.random.PRNGKey, input_shape: Tuple) -> Dict:
         raise NotImplementedError(f"init method has to be implemented for {self}")
 
+    def get_input_embeddings_key(self) -> str:
+        raise NotImplementedError(f"get_input_embeddings_key method has to be implemented for {self}")
+
+    def get_output_embeddings(self) -> Optional[jax.interpreters.xla._DeviceArray]:
+        """Overwrite this method for model classes that have output embeddings,
+        *e.g.* FlaxBertForMaskedLM
```
Comment (patrickvonplaten) [inline src/transformers/modeling_flax_utils.py:115]:
```suggestion
        *e.g.*, FlaxBertForMaskedLM.
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 5]
```diff
@@ -95,104 +95,44 @@
 """
 
 
-class FlaxBertLayerNorm(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/models/bert/modeling_flax_bert.py:98]: Let's use the official `flax.linen.LayerNorm` module to reduce future maintenance

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 6]
```diff
@@ -95,104 +95,44 @@
...
-        if self.scale:
-            mul = mul * jnp.asarray(self.weight)
-        y = (x - mean) * mul
-
-        if self.use_bias:
-            y = y + jnp.asarray(self.bias)
-        return y
-
-
-class FlaxBertEmbedding(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/models/bert/modeling_flax_bert.py:141]: Also let's use the official `flax.linen.Embed` module

File: tests/test_modeling_flax_common.py [hunk 7]
```diff
@@ -116,7 +122,46 @@ def test_equivalence_flax_pytorch(self):
                     len(fx_outputs_loaded), len(pt_outputs), "Output lengths differ between Flax and PyTorch"
                 )
                 for fx_output_loaded, pt_output in zip(fx_outputs_loaded, pt_outputs):
-                    self.assert_almost_equals(fx_output_loaded, pt_output.numpy(), 5e-3)
+                    self.assert_almost_equals(fx_output_loaded, pt_output.numpy(), 1e-3)
+
+    @is_pt_flax_cross_test
+    def test_equivalence_flax_to_pt(self):
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_common.py:128]: Test loading FLAX => PT now as well

File: tests/test_modeling_flax_common.py [hunk 8]
```diff
@@ -134,7 +179,7 @@ def test_from_pretrained_save_pretrained(self):
 
                 outputs_loaded = model_loaded(**prepared_inputs_dict)
                 for output_loaded, output in zip(outputs_loaded, outputs):
-                    self.assert_almost_equals(output_loaded, output, 5e-3)
+                    self.assert_almost_equals(output_loaded, output, 1e-3)
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_common.py:182]: Corrected bug allows tighter error bound

File: src/transformers/modeling_flax_utils.py [hunk 9]
```diff
@@ -330,13 +356,16 @@ def from_pretrained(
                     state = from_bytes(cls, state_f.read())
                 except UnpicklingError:
                     raise EnvironmentError(f"Unable to convert {archive_file} to Flax deserializable object. ")
+            # make sure all arrays are stored as jnp.arrays
+            state = unflatten_dict({k: jnp.asarray(v) for k, v in flatten_dict(state).items()})
```
Comment (patrickvonplaten) [inline src/transformers/modeling_flax_utils.py:360]: This is needed to ensure the loaded weights are in the correct format

**Review by sgugger — APPROVED** [review]
Great refactor, and nice to have the Flax to PT conversion. In the future it would be nice to add the TF to Flax and Flax to TF too (so that one does not have to go through PyTorch in between :-) )

File: src/transformers/modeling_flax_utils.py [hunk 10]
```diff
@@ -129,7 +142,20 @@ def params(self, params: Union[Dict, FrozenDict]):
...
+    def tie_weights(self):
+        output_embeddings = self.get_output_embeddings()
+
+        # tie word embedding with output embeddings
+        if output_embeddings is not None and self.config.tie_word_embeddings:
+            input_embbeddings_key = self.get_input_embeddings_key()
+            flattened_dict = flatten_dict(self.params)
+
+            # transpose weight and set correctly
+            flattened_dict[input_embbeddings_key] = jnp.asarray(output_embeddings.T)
```
[the remark this answers is not in the document]
Comment (CoderPat) [reply src/transformers/modeling_flax_utils.py:156]: Sorry for barging in, but I think this is correct, this will just initialized as tied but since jax arrays are immutable new copies will be created on gradient update that are not longer tied. The way they do it in the flax examples is to share a layer (see [this](https://github.com/google/flax/blob/a6282dba2f2be8d1724ea35f6e1d963097058cd9/examples/wmt/models.py#L342)), not sure how it would best integrate with the naming scheme used for pytorch

[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/modeling_flax_utils.py:156]: Yeah great catch! I realized this yesterday as well & [here](https://github.com/google/flax/discussions/1264#discussioncomment-644887) is a discussion here on the Flax repo regarding this point. The solution above is temporary so that at least during inference the model behaves correctly, *e.g.*, without this "hack" the following could be incorrect during inference:

```python 
bert = FlaxBertForMaskedLM.from_pretrained("some bert model that is tied", from_pt=True)
```

=> I'll add a big message stating the problem

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 11]
```diff
@@ -423,14 +364,19 @@ def __call__(self, hidden_states):
...
     config: BertConfig
     dtype: jnp.dtype = jnp.float32
+    bias_init: Callable[..., np.ndarray] = jax.nn.initializers.zeros
 
     def setup(self):
         self.transform = FlaxBertPredictionHeadTransform(self.config, dtype=self.dtype)
+        #        self.decoder = nn.Dense(self.config.vocab_size, dtype=self.dtype, use_bias=False)
         self.decoder = nn.Dense(self.config.vocab_size, dtype=self.dtype)
 
+    #        self.bias = self.param("bias", self.bias_init, (self.config.vocab_size,))
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/models/bert/modeling_flax_bert.py:374]: Sorry that was sloppy

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 12]
```diff
@@ -423,14 +363,22 @@ def __call__(self, hidden_states):
...
+        self.decoder = nn.Dense(self.config.vocab_size, dtype=self.dtype, use_bias=False)
+        self.bias = self.param("bias", self.bias_init, (self.config.vocab_size,))
 
-    def __call__(self, hidden_states):
+    def __call__(self, hidden_states, shared_embedding=None):
         hidden_states = self.transform(hidden_states)
-        hidden_states = self.decoder(hidden_states)
+
+        if shared_embedding is not None:
+            hidden_states = self.decoder.apply({"params": {"kernel": shared_embedding.T}}, hidden_states)
```
Comment (patrickvonplaten) [inline src/transformers/models/bert/modeling_flax_bert.py:377]: Ok, I think this could be the way to go for weight sharing - it's different from the PyTorch design where the weights are shared before the forward pass & therefore there is less code duplication in PyTorch as we can have a general function for it. 

If I understood correctly this is more difficult in Flax, so I think this is a good solution for now. It will mean that we'll re-write this code in all models that have weight sharing, but on the bright side it's also arguably less "magic" and the user has an easier time understanding that weights are shared when tracing a forward pass.

=> What do you think @sgugger @CoderPat @avital @marcvanzee ?

Comment (CoderPat) [reply src/transformers/models/bert/modeling_flax_bert.py:377]: It looks like a workable solution, however this will lead to creation of an unused kernel for the `self.decoder` right? Also why not have the shared embedding as a parameter rather than argument to the forward pass (similar to [the example that I mentioned](https://github.com/google/flax/blob/a6282dba2f2be8d1724ea35f6e1d963097058cd9/examples/wmt/models.py#L342))? Is it because it would screw up the names and therefor the automatic pytorch checkpoint conversion?

Comment (patrickvonplaten) [reply src/transformers/models/bert/modeling_flax_bert.py:377]: Yeah, I took this approach from the solution proposed [here](https://github.com/google/flax/discussions/1264#discussioncomment-644887) . The problem mostly is that we are quite restricted by the modularization structure to match PyTorch's ones. If I understand correctly, for your approach to work and for your approach to work I think we would have to create the layer outside of the word embedding module which doesn't work well with PyTorch's weight structure

Comment (patrickvonplaten) [reply src/transformers/models/bert/modeling_flax_bert.py:377]: Also, (if I understand correctly), there are no kernel weights created because the code actually never traces through the model at init -> so this should be fine in terms of memory

Comment (CoderPat) [reply src/transformers/models/bert/modeling_flax_bert.py:377]: Ok then I agree this seems to best way to do this. 
In the future if Flax adds something that allows weight sharing between more two `nn.module` maybe it might be worth for you guys to think about changing it
